adapter/service/parameters/service.go

```diff
@@ -0,0 +1,171 @@
+type Service interface {
+	adapter.Service
+
+	// SetParameter sets a parameter.
+	SetParameter(p Parameter) error
+	// SendParameterReport sends a parameter report. Returns true if the report was sent.
+	SendParameterReport(id string, force bool) (bool, error)
+	// SendSupportedParamsReport sends a supported parameters report. Returns true if the report was sent.
+	SendSupportedParamsReport(force bool) (bool, error)
+}
+
+// Config represents a service configuration.
+type Config struct {
+	Specification *fimptype.Service
+	Controller    Controller
+}
+
+// NewService creates new instance of a parameters FIMP service.
+func NewService(
+	publisher adapter.ServicePublisher,
+	cfg *Config,
+) Service {
+	cfg.Specification.Name = Parameters
+	cfg.Specification.EnsureInterfaces(requiredInterfaces()...)
+
+	return &service{
+		Service:           adapter.NewService(publisher, cfg.Specification),
+		controller:        cfg.Controller,
+		lock:              &sync.Mutex{},
+		reportingCache:    cache.NewReportingCache(),
+		reportingStrategy: cache.ReportOnChangeOnly(),
+	}
+}
+
+// service is a private implementation of a parameters FIMP service.
+type service struct {
+	adapter.Service
+
+	lock              *sync.Mutex
+	controller        Controller
+	reportingCache    cache.ReportingCache
+	reportingStrategy cache.ReportingStrategy
+}
+
+func (s *service) SetParameter(p Parameter) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if err := p.Validate(); err != nil {
+		return fmt.Errorf("%s: failed to validate parameter: %w", s.Name(), err)
+	}
+
+	spec, err := s.getParameterSpecification(p.ID)
+	if err != nil {
+		return fmt.Errorf("%s: failed to get parameter specification: %w", s.Name(), err)
+	}
+
+	if spec.ReadOnly {
+		return fmt.Errorf("%s: parameter is read only", s.Name())
+	}
+
+	if err = spec.ValidateParameter(p); err != nil {
+		return fmt.Errorf("%s: failed to validate parameter: %w", s.Name(), err)
+	}
+
+	if err = s.controller.SetParameter(p); err != nil {
+		return fmt.Errorf("%s: failed to set parameter ID %s: %w", s.Name(), p.ID, err)
+	}
+
+	return nil
+}
+
+func (s *service) SendParameterReport(id string, force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	parameter, err := s.controller.GetParameter(id)
+	if err != nil {
+		return false, fmt.Errorf("%s: failed to get parameter ID %s: %w", s.Name(), id, err)
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.reportingStrategy, EvtParamReport, "", parameter) {
+		return false, nil
+	}
+
+	message := fimpgo.NewObjectMessage(
+		EvtParamReport,
+		s.Name(),
+		parameter,
+		nil,
+		nil,
+		nil,
+	)
+	message.WithStorageStrategy(fimpgo.StorageStrategyAggregate, parameter.ID)
+
+	if err = s.SendMessage(message); err != nil {
+		return false, fmt.Errorf("%s: failed to send parameter report: %w", s.Name(), err)
+	}
+
+	return true, nil
+}
+
+func (s *service) SendSupportedParamsReport(force bool) (bool, error) {
```

Comment:
Maybe comments for methods...

Reply:
I don't have anything more to add than I've already wrote inside interface definition ;)